src/navigation/failure_identification/watchdog.py

```diff
@@ -1,11 +1,95 @@
 import numpy as np
 import rospy
 from pandas import DataFrame
+from util.ros_utils import get_rosparam
+from util.SO3 import SO3
+from typing import Tuple
+
+WINDOW_SIZE = get_rosparam("watchdog/window_size", 100)
+ANGULAR_THRESHOLD = get_rosparam("watchdog/angular_threshold", 0.001)
+LINEAR_THRESHOLD = get_rosparam("watchdog/linear_threshold", 0.55)
 
 
 class WatchDog:
-    def __init__(self):
-        pass
+    def get_start_end_positions(self, dataframe: DataFrame) -> Tuple[np.ndarray, np.ndarray]:
+        start_x, start_y, start_z = dataframe["x"].iloc[0], dataframe["y"].iloc[0], dataframe["z"].iloc[0]
+        start_pos = np.array([start_x, start_y, start_z])
+        end_x, end_y, end_z = dataframe["x"].iloc[-1], dataframe["y"].iloc[-1], dataframe["z"].iloc[-1]
+        end_pos = np.array([end_x, end_y, end_z])
+        return start_pos, end_pos
+
+    def get_start_end_rotations(self, dataframe: DataFrame) -> Tuple[SO3, SO3]:
+        start_rot = np.array(
+            [
+                dataframe["rot_x"].iloc[0],
+                dataframe["rot_y"].iloc[0],
+                dataframe["rot_z"].iloc[0],
+                dataframe["rot_w"].iloc[0],
+            ]
+        )
+        end_rot = np.array(
+            [
+                dataframe["rot_x"].iloc[-1],
+                dataframe["rot_y"].iloc[-1],
+                dataframe["rot_z"].iloc[-1],
+                dataframe["rot_w"].iloc[-1],
+            ]
+        )
+        return SO3(start_rot), SO3(end_rot)
+
+    def get_start_end_time(self, dataframe: DataFrame) -> Tuple[float, float]:
+        start_time = dataframe["time"].iloc[0]
+        end_time = dataframe["time"].iloc[-1]
+        return start_time, end_time
+
+    def check_angular_stuck(self, delta_time, delta_rot, dataframe) -> bool:
+        # check to make sure all cmd_vel_twist values are the same sign and cmd_vel_x values are 0
+        turn_are_all_same_sign = (
+            dataframe["cmd_vel_twist"].apply(np.sign).eq(np.sign(dataframe["cmd_vel_twist"].iloc[0])).all()
+        )
+        linear_are_all_zero = dataframe["cmd_vel_x"].eq(0).all()
+        print(f"turn all same sign: f{turn_are_all_same_sign}, linear_are_all_zero: f{linear_are_all_zero}")
+        if not turn_are_all_same_sign or not linear_are_all_zero:
```

Comment:
if all commands are zero, don't we want to return false?

Reply:
actually sorry, I guess what I'm trying to ask is will turn_are_all_the_same_sign return true if all the turn commands are 0? In which case you might report stuck despite not commanding a turn?

Reply:
so this is not actually possible since we only check if we're stuck when the state isn't off or done but I added an explicit check so its more flexible and also easier to understand since Riley brought up the same point